### ISSUE TITLE
gcc (g++): `ld: Assertion failed: (resultIndex < sectData.atoms.size())` (Xcode 15)

### ISSUE THREAD

**Issue by osalbahr** [Homebrew/homebrew-core#145991]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
$ brew config
HOMEBREW_VERSION: 4.1.13-4-gf1d345a
ORIGIN: https://github.com/Homebrew/brew
HEAD: f1d345a60e2c02f7e5c1b7930601b1921432cfb6
Last commit: 12 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 05f8a448ff23501ac834c6d9ce5802ecb23bf520
Core tap last commit: 13 minutes ago
Core tap branch: master
Core tap JSON: 26 Sep 00:21 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 4
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: quad-core 64-bit kabylake
Clang: 15.0.0 build 1500
Git: 2.42.0 => /usr/local/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 13.6-x86_64
CLT: 15.0.0.0.1.1694021235
Xcode: N/A
$ brew doctor
Your system is ready to brew.
```

### Verification

- [X] My "`brew doctor` output" says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.
- [X] I searched for recent similar issues at https://github.com/Homebrew/homebrew-core/issues?q=is%3Aissue and found no duplicates.

### What were you trying to do (and why)?

Compile c++ program for https://open.kattis.com/problems/mancala.

### What happened (include all command output)?

```console
$ g++-13 -std=gnu++17 minimal.cpp -o minimal
0  0x10d7cef43  __assert_rtn + 64
1  0x10d6d0f43  ld::AtomPlacement::findAtom(unsigned char, unsigned long long, ld::AtomPlacement::AtomLoc const*&, long long&) const + 1411
2  0x10d6ed431  ld::InputFiles::SliceParser::parseObjectFile(mach_o::Header const*) const + 19745
3  0x10d6fdb71  ld::InputFiles::parseAllFiles(void (ld::AtomFile const*) block_pointer)::$_7::operator()(unsigned long, ld::FileInfo const&) const + 657
4  0x7ff80075e066  _dispatch_client_callout2 + 8
5  0x7ff80077118f  _dispatch_apply_invoke_and_wait + 213
6  0x7ff800770692  _dispatch_apply_with_attr_f + 1207
7  0x7ff800770847  dispatch_apply + 45
8  0x10d796972  ld::AtomFileConsolidator::parseFiles(bool) + 370
9  0x10d71dd67  main + 12263
ld: Assertion failed: (resultIndex < sectData.atoms.size()), function findAtom, file Relocations.cpp, line 1336.
collect2: error: ld returned 1 exit status
$ cat minimal.cpp                                          
#include <iostream>
#include <vector>

using namespace std;

int main()
{
	vector<int> v;
	cout << v.size() << endl;
}
```

### What did you expect to happen?

```console
$ g++ -std=gnu++17 minimal.cpp -o minimal
$ ./minimal 
0
```

More info:
```console
$ g++-13 --version
g++-13 (Homebrew GCC 13.2.0) 13.2.0
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ g++ --version
Apple clang version 15.0.0 (clang-1500.0.40.1)
Target: x86_64-apple-darwin22.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
$ which g++-13
/usr/local/bin/g++-13
$ which g++
/usr/bin/g++
```

Afaik, this issue started after upgrading Xcode command line tools. I did some digging, and there's an Apple forum from 6 days ago:
https://developer.apple.com/forums/thread/737707

(also a Discussion https://discussions.apple.com/thread/255137447)

I am not sure if this is just a bug in Xcode, so I am opening this issue to track it.

See also:
- https://github.com/orgs/Homebrew/discussions/4794

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
g++-13 -std=gnu++17 minimal.cpp -o minimal
```

**Comment by Bo98** [comment]
This is indeed an Xcode bug.

Workaround is `-Wl,-ld_classic`. We may look into making that passed by default, if we can make it work in a backwards compatible fashion for those still on Xcode 14.

**Comment by osalbahr** [comment]
@Bo98 compiling with `-Wl,-ld_classic` seems to fix the issue. Are you suggesting rebuilding the `g++-*` bottles to add this flag by default?

```console
$ CXX='g++-13' make minimal
g++-13     minimal.cpp   -o minimal
0  0x1019aef43  __assert_rtn + 64
1  0x1018b0f43  ld::AtomPlacement::findAtom(unsigned char, unsigned long long, ld::AtomPlacement::AtomLoc const*&, long long&) const + 1411
2  0x1018cd431  ld::InputFiles::SliceParser::parseObjectFile(mach_o::Header const*) const + 19745
3  0x1018ddb71  ld::InputFiles::parseAllFiles(void (ld::AtomFile const*) block_pointer)::$_7::operator()(unsigned long, ld::FileInfo const&) const + 657
4  0x7ff80075e066  _dispatch_client_callout2 + 8
5  0x7ff80077118f  _dispatch_apply_invoke_and_wait + 213
6  0x7ff800770692  _dispatch_apply_with_attr_f + 1207
7  0x7ff800770847  dispatch_apply + 45
8  0x101976972  ld::AtomFileConsolidator::parseFiles(bool) + 370
9  0x1018fdd67  main + 12263
ld: Assertion failed: (resultIndex < sectData.atoms.size()), function findAtom, file Relocations.cpp, line 1336.
collect2: error: ld returned 1 exit status
make: *** [minimal] Error 1
$ CXX='g++-13' CXXFLAGS='-Wl,-ld_classic' make minimal
g++-13 -Wl,-ld_classic    minimal.cpp   -o minimal
$ ./minimal 
0
```

**Comment by osalbahr** [comment]
It seems that simply `brew reinstall -sv gcc` (it took `15391.05s user 2348.71s system 267% cpu 1:50:26.62 total`) fixed the issue for me, without modifying flags or anything:

```console
$ CXX='g++-13' make minimal
g++-13     minimal.cpp   -o minimal
$ ./minimal 
0
$ g++-13 --version
g++-13 (Homebrew GCC 13.2.0) 13.2.0
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ brew config
HOMEBREW_VERSION: 4.1.13-16-g39f76d6
ORIGIN: https://github.com/Homebrew/brew
HEAD: 39f76d635105aef18041f2eb1236945face1c6bf
Last commit: 9 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 05f8a448ff23501ac834c6d9ce5802ecb23bf520
Core tap last commit: 21 hours ago
Core tap branch: master
Core tap JSON: 26 Sep 18:43 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 4
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: quad-core 64-bit kabylake
Clang: 15.0.0 build 1500
Git: 2.42.0 => /usr/local/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 13.6-x86_64
CLT: 15.0.0.0.1.1694021235
Xcode: N/A
```

 Would it be possible to rebuild the `gcc` bottles(s) using the latest Xcode on Homebrew's CI, for testing purposes?

**Comment by Bo98** [comment]
Yeah, we already have a conditional where if Xcode 15 is detected then the build from source will use `ld-classic` as a linker

What we need to test however is whether that GCC built under Xcode 15 still works under Xcode 14. It might well do, but that needs to be confirmed.

**Comment by fxcoudert** [comment]
For the record, this appears fixes in the Xcode 15.1 beta

**Comment by osalbahr** [comment]
Unstale

**Comment by SMillerDev** [comment]
It's definitely stale, and fixed in the next XCode release. Nothing to do for brew.

**Comment by osalbahr** [comment]
> It's definitely stale, and fixed in the next XCode release. Nothing to do for brew.

Which release is it fixed in? The issue still persists in [Command Line Tools for Xcode 15.1 beta 3](https://developer.apple.com/download/all/?q=Command%20Line%20Tools%20for%20Xcode).

```console
% CXX='gcc-13' make minimal
gcc-13     minimal.cpp   -o minimal
ld: Undefined symbols:
  __ZNSolsEPFRSoS_E, referenced from:
      _main in cctbM3mR.o
  __ZNSolsEm, referenced from:
      _main in cctbM3mR.o
  __ZNSt8ios_base4InitC1Ev, referenced from:
      __Z41__static_initialization_and_destruction_0v in cctbM3mR.o
  __ZNSt8ios_base4InitD1Ev, referenced from:
      __Z41__static_initialization_and_destruction_0v in cctbM3mR.o
  __ZSt4cout, referenced from:
      _main in cctbM3mR.o
  __ZSt4endlIcSt11char_traitsIcEERSt13basic_ostreamIT_T0_ES6_, referenced from:
      _main in cctbM3mR.o
  __ZdlPvm, referenced from:
      __ZNSt15__new_allocatorIiE10deallocateEPim in cctbM3mR.o
  ___gxx_personality_v0, referenced from:
      /private/var/folders/d2/9dxqbydd5796hb7j623rk97w0000gn/T/cctbM3mR.o
collect2: error: ld returned 1 exit status
make: *** [minimal] Error 1
% brew config
HOMEBREW_VERSION: 4.1.22-57-g1130070
ORIGIN: https://github.com/Homebrew/brew
HEAD: 11300707964d85f9011bc8f7b418dfe23e1a8862
Last commit: 48 minutes ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: a85b622e488448c5ec12668f08a3a2d6e0170a86
Core tap last commit: 2 days ago
Core tap branch: bump-vulkan-headers-1.3.271
Core tap JSON: 01 Dec 14:00 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.m0f0vLisHl/org.xquartz:0
HOMEBREW_MAKE_JOBS: 4
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.1.4 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: quad-core 64-bit kabylake
Clang: 15.0.0 build 1500
Git: 2.43.0 => /usr/local/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 13.6.1-x86_64
CLT: 15.1.0.0.1.1699597245
Xcode: N/A
```

**Comment by Bo98** [comment]
That doesn't seem to be the same issue. Is that from the same example code as the original report?

**Comment by fxcoudert** [comment]
This issue is about the bug `ld: Assertion failed: (resultIndex < sectData.atoms.size())` in Apple linker. That is confirmed fixed in Xcode 15.1 beta 3 (and CLT).

Your latest error message seems to be because libstdc++ is missing, probably because `CXX` should be `g++-13` instead of `gcc-13`.
